Fix #19792: raise error when tab RTE components are empty and prevent saving

alice21mota opened the pull request:
## Overview

<!--
READ ME FIRST:
Please answer *all* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes #19792 .
2. This PR does the following: prevents user from clicking "Done" when at least one of the tab components is empty. Displays an error message indicating the error.
3. (For bug-fixing PRs only) The original bug occurred because: There was no validation for the tab RTE components, which would generate errors when saving empty components.

## Essential Checklist

Please follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).

- [x] I have linked the issue that this PR fixes in the "Development" section of the sidebar.
- [x] I have checked the "Files Changed" tab and confirmed that the changes are what I want to make.
- [x] I have written tests for my code.
- [x] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...", followed by a short, clear summary of the changes.
- [x] I have assigned the correct reviewers to this PR (or will leave a comment with the phrase "@{{reviewer_username}} PTAL" if I can't assign them directly).

## Proof that changes are correct

https://github.com/oppia/oppia/assets/100164400/220303e9-d54d-4047-840c-ae8cd0e37d2d

## PR Pointers

- Never force push! If you do, your PR will be closed.
- To reply to reviewers, follow these instructions: https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

alice21mota committed:
Fix: Handle empty content error in tab interaction

Prevents user from saving when a component of a tab is empty, displaying an error message

alice21mota committed:
fix: linting errors and unnecessary code removed

alice21mota commented:
@seanlip PTAL

seanlip commented on core/templates/services/rte-helper-modal.controller.ts:
```diff
@@ -253,6 +253,18 @@ export class RteHelperModalComponent {
       }
+    } else if (this.componentId === this.COMPONENT_ID_TABS) {
+      // Value[0] corresponds to tab Contents.
+      for (let tab = 0; tab < value[0].length; tab++) {
```
This is a number, so it should be tabIndex rather than tab

alice21mota replied:
Changes made! Thank you !

seanlip commented on core/templates/services/rte-helper-modal.controller.ts:
```diff
@@ -253,6 +253,18 @@ export class RteHelperModalComponent {
+      for (let tab = 0; tab < value[0].length; tab++) {
+        if (value[0][tab].title === '' || value[0][tab].content === '') {
+          this.updateRteErrorMessage(
+            'Please ensure that each title and content are filled'
```
Try to use a more specific message (e.g. "Please ensure that the title of tab 2 is filled in") so that the creator knows what exactly to do without hunting around.

alice21mota replied:
Changes made, thank you !

seanlip commented on core/templates/services/rte-helper-modal.controller.ts:
```diff
@@ -253,6 +253,18 @@ export class RteHelperModalComponent {
         return;
       }
+    } else if (this.componentId === this.COMPONENT_ID_TABS) {
+      // Value[0] corresponds to tab Contents.
```
What is "tab Contents"? There are multiple tabs, so not sure what you mean -- is it the contents for all tabs or just the first one? Perhaps worth clarifying your comment here.

alice21mota replied:
I altered the comment, yes, it is the contents for all tabs. I believe it makes more sense now. Thank you !

seanlip reviewed (changes requested):
Thanks @alice21mota! Left some notes.

alice21mota committed:
fix: changes request by reviwer
 change comment related to Value[0] in line 257 to be more specific
 replace tab with tabIndex
 make error messages more explicit for users
 fix tests to accomodate changes

alice21mota committed:
fix: changes request by reviwer
dquote> change comment related to Value[0] in line 257 to be more specific
dquote> replace tab with tabIndex
dquote> make error messages more explicit for users
dquote> fix tests to accomodate changes

alice21mota committed:
fix: changes request by reviwer
dquote> change comment related to Value[0] in line 257 to be more specific
dquote> replace tab with tabIndex
dquote> make error messages more explicit for users
dquote> fix tests to accomodate changes

alice21mota commented:
Hi @seanlip PTAL!
I made the changes you requested, let me know if there are any remaining issues.
This is how it would appear now in case of empty component:

https://github.com/oppia/oppia/assets/100164400/9e58904e-24ba-466a-b2c2-2ae092fa9099

seanlip commented on core/templates/services/rte-helper-modal.controller.spec.ts:
```diff
@@ -356,21 +356,22 @@ describe('RteHelperModalComponent', () => {
       expect(component.isErrorMessageNonempty()).toBe(true);
       expect(component.errorMessage).toBe(
-        'Please ensure that each title and content are filled'
+        'Please ensure that tab 1 title is filled.'
```
Please ensure that the title of tab 1 is filled

alice21mota replied:
Done!

seanlip commented on core/templates/services/rte-helper-modal.controller.spec.ts:
```diff
@@ -356,21 +356,22 @@ describe('RteHelperModalComponent', () => {
       expect(component.isErrorMessageNonempty()).toBe(true);
       expect(component.errorMessage).toBe(
-        'Please ensure that each title and content are filled'
+        'Please ensure that tab 2 content is filled.'
```
Please ensure that the content of tab 2 is filled

alice21mota replied:
Done!

seanlip commented on core/templates/services/rte-helper-modal.controller.ts:
```diff
@@ -254,11 +254,16 @@ export class RteHelperModalComponent {
-      // Value[0] corresponds to tab Contents.
-      for (let tab = 0; tab < value[0].length; tab++) {
-        if (value[0][tab].title === '' || value[0][tab].content === '') {
+      // Value[0] corresponds to all tabs Contents and titles.
```
all tab contents and titles.

(tabs --> tab; do not capitalize "Contents")

alice21mota replied:
Done!

seanlip reviewed (changes requested):
Thanks @alice21mota. I think this looks good, just a few minor notes and it should be good to merge. Sorry for the delay, the past few days have been quite hectic.

Also, just a note -- please don't resolve review comments, let the reviewer do that. Otherwise the reviewer just has to open everything up again to check that the comments are addressed. Thanks!

alice21mota commented:
Thank you for the notes @seanlip ! Will address and commit them shortly.

alice21mota commented:
Hi @seanlip ! Is it possible to re-run the checks? Many of them were cancelled after one failed, but I don´t believe it has anything to do with the changes I made. If not how should I proceed?

seanlip commented:
@alice21mota Yup, this is a known flake (#19732). I'll re-run, thanks!

In future, if you see an error, please search for it on GitHub and see if it's filed already. If so then reference it when asking for a re-run; if not, then either the issue is with your PR or a new issue should be filed. Thanks!

alice21mota commented:
Thank you @seanlip ! I believe the same thing occurred, not sure why. Will add the info to the existing issue.

seanlip commented:
@alice21mota The new error is different, please follow the procedure I outlined above, thanks.

alice21mota committed:
fix: displayed errors messages
alter displayed messages when a tab component is not filled
change corresponding tests accordingly
fix comment typos

alice21mota commented:
Hi @seanlip PTAL!
I made the requested changes. Hope it is all good now. Awaiting your feedback and thanks in advance.
In relation to one of the failed e2e tests, I believe it is related to the issue [#17279](https://github.com/oppia/oppia/issues/17279), and therefore a flake. As for the other failed e2e test, e believe it may be due to some network connection error, but correct me if I am wrong.

seanlip reviewed (approved):
It looks good, thanks @alice21mota. Let's try to rerun this once https://github.com/oppia/oppia/pull/20287 is merged.

One note, can you please respond to the existing review comments (with "Done" if you've done them)? Let's follow the proper review process. Thanks!

alice21mota commented:
Done! Thank you @seanlip !

seanlip committed:
Merge branch 'develop' into validation-tabs-rte-component

seanlip committed:
Merge branch 'develop' into validation-tabs-rte-component

alice21mota commented:
Thank you @seanlip ! Will do :)
